library/std/unstable_arithmetic.qs

```diff
@@ -512,140 +512,92 @@ namespace Microsoft.Quantum.Unstable.Arithmetic {
+    //   if c < x { action(target) }  | ApplyIfLessL
+    //   if c <= x { action(target) } | ApplyIfLessOrEqualL
+    //   if c == x { action(target) } | ApplyIfEqualL
+    //   if c >= x { action(target) } | ApplyIfGreaterOrEqualL
+    //   if c > x { action(target) }  | ApplyIfGreaterL
+    //
+    // Compare two qubit registers in a little-endian format and apply action
+    //   if x < y { action(target) }  | ApplyIfLessLE
+    //   if x <= y { action(target) } | ApplyIfLessOrEqualLE
+    //   if x == y { action(target) } | ApplyIfEqualLE
+    //   if x >= y { action(target) } | ApplyIfGreaterOrEqualLE
+    //   if x > y { action(target) }  | ApplyIfGreaterLE
     //
 
     /// # Summary
-    /// Computes `target ^= (c < x)`, that is, inverts `target`
+    /// Computes `if (c < x) { action(target) }`, that is, applies `action` to `target`
     /// if a BigInt value `c` is less than the little-endian qubit register `x`
-    operation InvertIfLessL (c : BigInt, x : Qubit[], target : Qubit) : Unit is Adj + Ctl {
-        ApplyActionIfGreaterThanOrEqualConstant(false, X, c + 1L, x, target);
+    operation ApplyIfLessL<'T> (
+        action : 'T => Unit is Adj + Ctl,
+        c : BigInt,
+        x : Qubit[],
+        target : 'T) : Unit is Adj + Ctl {
+
+        ApplyActionIfGreaterThanOrEqualConstant(false, action, c + 1L, x, target);
     }
 
     /// # Summary
-    /// Computes `target ^= (c <= x)`, that is, inverts `target`
+    /// Computes `if (c <= x) { action(target) }`, that is, applies `action` to `target`
     /// if a BigInt value `c` is less or equal to the little-endian qubit register `x`
-    operation InvertIfLessOrEqualL (c : BigInt, x : Qubit[], target : Qubit) : Unit is Adj + Ctl {
-        ApplyActionIfGreaterThanOrEqualConstant(false, X, c, x, target);
+    operation ApplyIfLessOrEqualL<'T> (
+        action : 'T => Unit is Adj + Ctl,
+        c : BigInt,
+        x : Qubit[],
+        target : 'T) : Unit is Adj + Ctl {
+
+        ApplyActionIfGreaterThanOrEqualConstant(false, action, c, x, target);
     }
 
     /// # Summary
-    /// Computes `target ^= (c == x)`, that is, inverts `target`
+    /// Computes `if (c == x) { action(target) }`, that is, applies `action` to `target`
     /// if a BigInt value `c` is equal to the little-endian qubit register `x`
-    operation InvertIfEqualL (c : BigInt, xs : Qubit[], target : Qubit) : Unit is Adj + Ctl {
+    operation ApplyIfEqualL<'T> (
+        action : 'T => Unit is Adj + Ctl,
+        c : BigInt,
+        xs : Qubit[],
+        target : 'T) : Unit is Adj + Ctl {
+
         let cBitSize = BitSizeL(c);
         let xLen = Length(xs);
         if (cBitSize <= xLen) {
             let bits = BigIntAsBoolArray(c, Length(xs));
             within {
                 ApplyPauliFromBitString(PauliX, false, bits, xs);
             } apply {
-                Controlled X(xs, target);
+                Controlled action(xs, target);
```

Comment:
I suggest to use AndChain here and only singly-controlled the action.

Reply:
I would prefer not to do this, at least for now. I can see that all other (not equality) comparisons result in a singly controlled action. But if I add the And chain here, user won't be able to avoid it. On the other hand, if I do not add it here, user will be able to provide it if needed.
I'm even thinking we could provide a generic operation that would turn controllable operation into a singly-controlled operation.

Reply:
I gave it a try and I updated code to apply AND chain. It is faster in simulation and should be better for resource estimation.